cilium/socket_option_source_address.cc

```diff
@@ -77,15 +80,23 @@ bool SourceAddressSocketOption::setOption(
     return true;
   }
 
-  if (source_address->ip() && dest_fs_->getDestinationAddress() &&
-      dest_fs_->getDestinationAddress()->ip() &&
-      source_address->ip()->addressAsString() ==
-          dest_fs_->getDestinationAddress()->ip()->addressAsString()) {
-    ENVOY_LOG(trace,
-              "Skipping restore of local address on socket: {} - source address is same as "
-              "destination address {}",
-              socket.ioHandle().fdDoNotUse(), source_address->ip()->addressAsString());
-    return true;
+  if (source_address->ip() && dest_fs_ && dest_fs_->getDestinationAddress() &&
+      dest_fs_->getDestinationAddress()->ip()) {
+    const auto& dst_addr = dest_fs_->getDestinationAddress()->ip()->addressAsString();
+    if (source_address->ip()->addressAsString() == dst_addr) {
+      ENVOY_LOG(trace,
+                "Skipping restore of local address on socket: {} - source address is same as "
+                "destination address {}",
+                socket.ioHandle().fdDoNotUse(), dst_addr);
+      return true;
+    }
+    // Also skip using the original source address if destination is local.
+    // Local destinations have a policy configured on their address.
+    if (policy_fs_ && policy_fs_->exists(dst_addr)) {
+      ENVOY_LOG(trace, "Skipping restore of local address on socket: {} - destination is local {}",
+                socket.ioHandle().fdDoNotUse(), dst_addr);
```

Comment:
is there any chance the dest_addr is different with source_address i.e. 2 different pods on the same pod ?

Reply:
We have a policy entry keyed with each address of the (local) endpoint (e.g., IPv4 and IPv6). I don't think we currently can have two IPv4 addresses or two IPv6 addresses on any given endpoint, but if we do in the future, then we'd have to make sure each of them is reported in the NPDS update for the endpoint. This means that for every local endpoint IP there should be a policy configured, and we can use that fact to distinguish which endpoints are local and which are not, like here. Furthermore, we avoid using the original source address here in order to avoid 5-tuple collisions in the host networking namespace in case the destination pod happens to (also) have a HTTP policy (Envoy redirection). That can only happen if there is a policy for the destination endpoint.